runtime/compiler/x/amd64/runtime/AMD64CompressString_nasm.inc

```diff
@@ -0,0 +1,235 @@
+		mov  QWORD  [rdi], rax			; write the low 2 bytes of the 4 byte value in the destination
+
+		mov  rax, QWORD  [rsi+16]  		; load 4 bytes from the source array into EAX (2 chars)
+		mov  rbx, QWORD  [rsi+24]  		; load 4 bytes from the source array into EAX (2 chars)
+		or   rdx, rax				;
+		or   rdx, rbx
+		shl  rbx, 8					; shift right by 8
+		or   rax, rbx					; or the 2 values
+		mov  QWORD  [rdi+8], rax			; write the low 2 bytes of the 4 byte value in the destination
+
+		add  rdi, 16
+		add  rsi, 32		
+		loop eightcharsJ
+                mov  rax, rdx
+                shl  rdx, 32
+                or   rax, rdx
+		shr  rax, 32
+                ret
+;_compressStringJ endp
+;
+
+_andORString: ;    PROC
+		shr  rcx, 4
+		add  rsi, rax
+		add  rsi, rax
+		mov  rbx, 0
+		mov  rdx, 0ffffffffffffffffh
+eightchars2:
+		mov  rax, QWORD  [rsi]  		; load 4 bytes from the source array into EAX (2 chars)
+		or   rbx, rax
+		and  rdx, rax
+		mov  rax, QWORD  [rsi+8]  		; load 4 bytes from the source array into EAX (2 chars)
+		or   rbx, rax
+		and  rdx, rax
+		mov  rax, QWORD  [rsi+16]  		; load 4 bytes from the source array into EAX (2 chars)
+		or   rbx, rax
+		and  rdx, rax
+		mov  rax, QWORD  [rsi+24]  		; load 4 bytes from the source array into EAX (2 chars)
+		or   rbx, rax
+		and  rdx, rax
+		add  rsi, 32		
+		loop eightchars2
+
+		mov  rax, rbx                           ; building the or
+		mov  rcx, rdx                           ; building the and
+		shr  rbx, 32                            ; building the or
+		shr  rdx, 32                            ; building the and
+		or   ebx, eax				; building the or
+		and  edx, ecx				; building the and
+		mov  eax, ebx				; building the or
+		mov  ecx, edx				; building the and
+		shr  ebx, 16				; building the or
+		shl  edx, 16				; building the and
+		or   bx, ax				; building the or
+		and  edx, ecx				; building the and
+
+		mov  dx, bx
+                ret
+;_andORString endp
+;
+
+
+_compressStringNoCheckJ: ;    PROC
+		shr  rcx, 4
+		add  rsi, rax
+		add  rsi, rax
+eightcharsNoCheckJ:
+		mov  rax, QWORD  [rsi]  		; load 4 bytes from the source array into EAX (2 chars)
+		mov  rbx, QWORD  [rsi+8]  		; load 4 bytes from the source array into EAX (2 chars)
+		shl  rbx, 8					; shift right by 8
+		or   rax, rbx					; or the 2 values
+		mov  QWORD  [rdi], rax			; write the low 2 bytes of the 4 byte value in the destination
+
+		mov  rax, QWORD  [rsi+16]  		; load 4 bytes from the source array into EAX (2 chars)
+		mov  rbx, QWORD  [rsi+24]  		; load 4 bytes from the source array into EAX (2 chars)
+		shl  rbx, 8					; shift right by 8
+		or   rax, rbx					; or the 2 values
+		mov  QWORD  [rdi+8], rax			; write the low 2 bytes of the 4 byte value in the destination
+
+		add  rdi, 16
+		add  rsi, 32		
+		loop eightcharsNoCheckJ
+                ret
+;_compressStringNoCheckJ endp
+;
+
+
+_compressStringNoCheck: ; PROC 
+		shr  rcx, 4
+		add  rsi, rax
+		add  rsi, rax
```

Comment:
We use tabs in this file but spaces everywhere else. Is this intended?

Reply:
This tabs vs. space inconsistency is due to the fact that for some of the files I've used a heavily modified version of the masm2gas script to do some initial conversion, while others were done without the help of the script. I am working on the formatting and will push an update while it is being reviewed.